torch/_dynamo/output_graph.py

```diff
@@ -1256,6 +1257,15 @@ def compile_and_call_fx_graph(self, tx, rv, root):
         """
         from .decorators import disable
 
+        if (ds := tx.distributed_state) is not None and ds.all_states is None:
```

Comment:
basic lifecycle question (i'm mainly trying to understand why we need to check if `ds.all_states` has already been set here) - I see that `tx.distributed_state is initialized in `convert_frame._compile()`, and set the first time `compile_and_call_fx_graph` is called.

Does that mean:
(1) we will have a unique `distributed_state` for every frame + recompile ([X/Y] in tlparse
(2) `compile_and_call_fx_graph` can be called multiple times due to dynamo restarting analysis (e.g. `[X/Y_n]`), so if we had already done the collective we don't want to re-do it here

Reply:
distributed_state is shared across all restarts. The inner all_states is empty until we've run the compiler collective that populates it.

Once we've done the collective once, we no longer redo it. Actually, it should be impossible to restart analysis after the compiler collective, but I haven't asserted this.